staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication_test.go

```diff
@@ -18,15 +18,81 @@ package filters
 
 import (
 	"errors"
-	"net/http"
-	"net/http/httptest"
-	"testing"
-
 	"k8s.io/apiserver/pkg/authentication/authenticator"
 	"k8s.io/apiserver/pkg/authentication/user"
 	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"net/http"
+	"net/http/httptest"
+	"testing"
 )
 
+func TestAuthenticateRequestWithAud(t *testing.T) {
+	success := make(chan int, 1)
+	failed := make(chan int, 1)
+	testcases := []struct {
+		name          string
+		apiAuds       []string
+		respAuds      []string
+		expectSuccess bool
+	}{
+		{
+			name:          "audience in response",
+			apiAuds:       nil,
+			respAuds:      []string{"other"},
+			expectSuccess: true,
+		},
+		{
+			name:          "with api audience",
+			apiAuds:       authenticator.Audiences([]string{"other"}),
+			respAuds:      nil,
+			expectSuccess: true,
+		},
+		{
+			name:          "api audience matching response audience",
+			apiAuds:       authenticator.Audiences([]string{"other"}),
+			respAuds:      []string{"other"},
+			expectSuccess: true,
+		},
+		{
+			name:          "api audience non-matching response audience",
+			apiAuds:       authenticator.Audiences([]string{"other"}),
+			respAuds:      []string{"some"},
+			expectSuccess: false,
+		},
+	}
+	for _, tc := range testcases {
+		auth := WithAuthentication(
+			http.HandlerFunc(func(_ http.ResponseWriter, req *http.Request) {
+				if tc.expectSuccess {
+					success <- 1
+				} else {
+					t.Errorf("unexpected call to handler")
+				}
+			}),
+			authenticator.RequestFunc(func(req *http.Request) (*authenticator.Response, bool, error) {
+				if req.Header.Get("Authorization") == "Something" {
+					return &authenticator.Response{User: &user.DefaultInfo{Name: "user"}, Audiences: authenticator.Audiences(tc.respAuds)}, true, nil
+				}
+				return nil, false, errors.New("Authorization header is missing.")
+			}),
+			http.HandlerFunc(func(_ http.ResponseWriter, _ *http.Request) {
+				if tc.expectSuccess {
+					t.Errorf("unexpected call to failed")
+				} else {
+					failed <- 2
+				}
+			}),
+			tc.apiAuds,
+		)
+		auth.ServeHTTP(httptest.NewRecorder(), &http.Request{Header: map[string][]string{"Authorization": {"Something"}}})
+		if tc.expectSuccess {
+			_ = <-success
```

Comment:
on failure, this just blocks, rather than give a helpful message. suggest setting up `success, failure := 0, 0` inside each testcase loop, incrementing that from within the handlers, then checking that success==1 && failure==0 in success cases, and success==0 && failure==1 in failure cases